From e5577b7478924d89cfd0daeeb4b6443a8560934f Mon Sep 17 00:00:00 2001
From: Hans Boehm <hboehm@google.com>
Date: Fri, 29 Oct 2021 14:57:37 -0700
Subject: [PATCH 09/13] [master] Assign lower level to Small IRT table lock

It can be acquired after Class loader classes lock. Jvmti
GetLoadedClasses traverses classes with a visitor that may need
to allocate a LocalRef.

Bug: 184847225
Test: Build and boot AOSP
Change-Id: Ic19b2a3238321bc21440064f0d1b531ed254ea35
---
 runtime/indirect_reference_table.cc | 2 +-
 runtime/indirect_reference_table.h  | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/runtime/indirect_reference_table.cc b/runtime/indirect_reference_table.cc
index 4a06b2f2d7..fbe678bea2 100644
--- a/runtime/indirect_reference_table.cc
+++ b/runtime/indirect_reference_table.cc
@@ -80,7 +80,7 @@ static inline MemMap NewIRTMap(size_t table_bytes, std::string* error_msg) {
 }
 
 SmallIrtAllocator::SmallIrtAllocator()
-    : small_irt_freelist_(nullptr), lock_("Small IRT table lock") {
+    : small_irt_freelist_(nullptr), lock_("Small IRT table lock", LockLevel::kGenericBottomLock) {
 }
 
 // Allocate an IRT table for kSmallIrtEntries.
diff --git a/runtime/indirect_reference_table.h b/runtime/indirect_reference_table.h
index 6bdf315143..4433e9eeca 100644
--- a/runtime/indirect_reference_table.h
+++ b/runtime/indirect_reference_table.h
@@ -247,7 +247,7 @@ class SmallIrtAllocator {
   // Repository of MemMaps used for small IRT tables.
   std::vector<MemMap> shared_irt_maps_;
 
-  Mutex lock_;
+  Mutex lock_;  // Level kGenericBottomLock; acquired before mem_map_lock_, which is a C++ mutex.
 };
 
 class IndirectReferenceTable {
-- 
2.34.0

